### PR TITLE
Fixed performance regression in calling procedures

### PR DESCRIPTION
**Pull request by craigtaverner**

### CODE REVIEW

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/procs/ProcedureExecutionResult.scala [hunk 1]
```diff
@@ -95,9 +95,8 @@ class ProcedureExecutionResult(context: QueryContext,
   override def accept[EX <: Exception](visitor: QueryResultVisitor[EX]): Unit = {
     executionResults.foreach { res =>
       val fieldArray = new Array[AnyValue](indexResultNameMappings.size)
-      var i = 0
-      for ((pos, _, typ) <- indexResultNameMappings) {
-        fieldArray(i) = typ match {
+      for (pos <- indexResultNameMappings.indices) {
+        fieldArray(pos) = indexResultNameMappings(pos)._3 match {
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/procs/ProcedureExecutionResult.scala]: is this a bottleneck? If not I think the old way looks nicer, not a fan of `._3`

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala [hunk 2]
```diff
@@ -98,9 +99,13 @@ abstract class StandardInternalExecutionResult(context: QueryContext, runtime: R
   override def accept[E <: Exception](visitor: ResultVisitor[E]): Unit = {
     accept(new QueryResultVisitor[E] {
       override def visit(record: QueryResult.Record): Boolean = {
-        val row = new MapBasedRow
-        row.map = fieldNames().zip(record.fields().map(context.asObject)).toMap
-        visitor.visit(row)
+        val names = fieldNames()
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala]: Probably not super important but `val names = fieldNames() ` can be done outside `visit`

**Review by pontusmelke — APPROVED** [review]
Nice, had some comments that are not terribly important.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/procs/ProcedureExecutionResult.scala (hunk 1 above)
Comment (craigtaverner) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/procs/ProcedureExecutionResult.scala]: Amazingly the profiling showed this to be 2x faster for this part of the profile. It turned out that the call stack created by scala for the tuple foreach was much, much deeper than the one for the int range foreach. No idea why, but the performance change was very real.

So we did not do this whimsically.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala (hunk 2 above)
Comment (craigtaverner) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala]: And I wonder if the fact that fieldNames is now a `val` would get this optimised by the compiler the same in all cases?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala [hunk 3]
```diff
@@ -97,10 +97,14 @@ abstract class StandardInternalExecutionResult(context: QueryContext, runtime: R
 
   override def accept[E <: Exception](visitor: ResultVisitor[E]): Unit = {
     accept(new QueryResultVisitor[E] {
+      val names = fieldNames()
       override def visit(record: QueryResult.Record): Boolean = {
-        val row = new MapBasedRow
-        row.map = fieldNames().zip(record.fields().map(context.asObject)).toMap
-        visitor.visit(row)
+        val fields = record.fields()
+        val mapData = new mutable.AnyRefMap[String, Any](names.length)
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala:103]: We do the same thing in `PipeExecutionResult`, it should be just as inefficient there. Can we apply the same fix there?

Comment (craigtaverner) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/executionplan/StandardInternalExecutionResult.scala:103]: Good point. I've added a new commit for this. I also profiled it with `MATCH (n) RETURN n` and in that case the previous code used 40% of the result generation time to convert the results to external types with this method, while after the commit the code uses 20% of the time. So, as before, we double the performance of this specific algorithm.

**Pull request merged**